### PR TITLE
Feature model and associations

### PR DESCRIPTION
**Pull request by UbdaNam**
## What does this PR do?

This PR is raised to implement the following features: 👍 :

- [x] Add models and associations
- [x] Create custom methods in the models
- [x] Use GitFlow to introduce new features
- [x] Setup rubocop and stylelint linters

## Description of the tasks completed 👇 

- [x] Add models and associations
- [x] Install rails gem
- [x] Use the linter tool that analyzes my source code to flag programming errors, bugs, stylistic errors, and suspicious constructs. 💯 
  
## How should it be tested

Clone the repository
```
git clone https://github.com/UbdaNam/Blogiverse.git
```
Navigate to the folder
```
cd Blogiverse
```
Checkout the branch
```
git checkout feature_model_and_associations
```
Run migrations
```
rails db:migrate
```
Test custom modal methods
```
rails c
```

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: :author_id, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :likes, dependent: :destroy
```
Comment (MahmoudBakr23) [inline app/models/user.rb:4]: - [x] Please add the `foreign_key` attribute to the comments and likes associations as well to ensure that the owner of each record in those models is a User but referred to as an Author as per requirements. 🙏

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  has_many :likes, dependent: :destroy
+
+  def update_posts_count
+    author.update(posts_counter: author.posts.count)
+  end
+
+  def recent_comments
+    comments.last(5)
+  end
+end
```
Comment (MahmoudBakr23) [inline app/models/post.rb]: OPTIONAL
- You can also consider using the helper method `after_save` to ensure that with every update related to this specific model `Post` then invoke the `update_posts_count` method immediately. 👍

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post, counter_cache: :likes_counter
+
+  def update_post_like_counter
+    post.update(likes_counter: post.likes.count)
+  end
+end
```
Comment (MahmoudBakr23) [inline app/models/like.rb]: OPTIONAL
- You can also consider using the helper method `after_save` to ensure that with every update related to this specific model `Like` then invoke the `update_post_like_counter` method immediately. 👍

File: app/models/comment.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post, counter_cache: :comments_counter
+
+  def update_post_counter
+    post.update(comments_counter: post.comments.count)
+  end
+end
```
Comment (MahmoudBakr23) [inline app/models/comment.rb]: OPTIONAL
- You can also consider using the helper method `after_save` to ensure that with every update related to this specific model `Comment` then invoke the `update_post_counter` [Please consider renaming this method to update_comment_counter for more clarity and readability] method immediately. 👍

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
Comment (MahmoudBakr23) [inline app/models/comment.rb]: - [x] As per requirements, please modify the association to belong to `author` with a class_name attribute set to `User`

File: app/models/like.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (MahmoudBakr23) [inline app/models/like.rb]: - [x] As per requirements, please modify the association to belong to `author` with a class_name attribute set to `User`

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
Hello @UbdaNam 😃, 

Great work and effort you put into this project so far.

## Highlights 🚀 

There are some excellent points that you have made perfectly. Keep up! 💪
- Well done so far working with the associations of your models
- Descriptive PR and README
- Correct Linters

However, there is always room for improvement.

> I'd like to remind you that your long work hours are appreciated. However, the code review process is for improving your code, not criticizing it. 😄

 ## Required Changes ♻️
_Please check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_See the comments under the review._

Cheers, and Happy coding!👏👨‍💻💻

Feel free to mention @MahmoudBakr23 to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (UbdaNam) [reply app/models/post.rb]: hi @MahmoudBakr23 thank you for your suggestions actually, I was adding it for requirement purposes, I was actually caching the counter in every update so the method is unneccessary

File: app/models/comment.rb (hunk 4 above)
Comment (UbdaNam) [reply app/models/comment.rb]: hi @MahmoudBakr23 thank you for your suggestions actually, I was adding it for requirement purposes, I was actually caching the counter in every update so the method is unnecessary

File: app/models/user.rb (hunk 1 above)
Comment (peterrobert) [reply app/models/user.rb:4]: This change has been done well

File: app/models/post.rb (hunk 2 above)
Comment (peterrobert) [reply app/models/post.rb]: This change has been done well

File: app/models/like.rb (hunk 3 above)
Comment (peterrobert) [reply app/models/like.rb]: This change has been done well

File: app/models/comment.rb (hunk 5 above)
Comment (peterrobert) [reply app/models/comment.rb]: This change has been done well

File: app/models/like.rb (hunk 6 above)
Comment (peterrobert) [reply app/models/like.rb]: this change has been done well

**Review by peterrobert — APPROVED** [review]
Hi @UbdaNam, 


### Highlights!

[ ] Good feature branch name ✔️
[ ] Descriptive pull request ✔️


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by UbdaNam** [comment]
Thank you @peterrobert for reviewing my PR

**Pull request merged**